qiskit/circuit/bit.py

```diff
@@ -56,11 +57,17 @@ def __getitem__(self, item):
             raise IndexError
 
     def __hash__(self):
-        return hash((self.register, self.index))
+        if self._hash is None:
+            self._hash = hash((self.register, self.index))
+        return self._hash
 
     def __eq__(self, other):
         if isinstance(other, Bit):
-            return other.index == self.index and other.register == self.register
+            return other.index == self.index and \
+                   other.register == self.register and \
+                   type(self) is type(other)
         if isinstance(other, tuple):
+            warn('Equality check between a tuple and a Bit instances is deprecated. '
```

Comment:
i don't think we need this warning because Bit was never on stable.

Reply:
this warning triggers when `new_bit == old_bit`, If the user is still using old bits.